File: hera/bot.py
"""Bot wiring: shared services, background market ticker and event dispatch."""

from __future__ import annotations

import logging

import discord
from discord import app_commands
from discord.ext import commands, tasks

from .config import config
from .database import Database
from .errors import HeraError
from .formatting import money, percent, price, signed
from .services.economy import EconomyService
from .services.market import MarketService
from .services.trading import TradingService

log = logging.getLogger("hera")


def _usage_error(ctx: commands.Context, error: commands.CommandError) -> str:
    """Explain a mistyped command and show how to call it correctly."""
    if isinstance(error, commands.MissingRequiredArgument):
        return f"`{error.param.name}` is required: {usage_for(ctx)}"
    if isinstance(error, commands.BadArgument):
        return f"{error}\n{usage_for(ctx)}"
    return f"That is not how the command is used: {usage_for(ctx)}"


def usage_for(ctx: commands.Context) -> str:
    """Render a copy-pasteable usage line for the invoked command."""
    command = ctx.command
    if command is None:
        return f"Use `{config.prefix}help`."
    return f"`{config.prefix}{command.qualified_name} {command.signature}`".rstrip()


class HeraBot(commands.Bot):
    """A bot instance that owns the database and every shared service."""

    def __init__(self) -> None:
        intents = discord.Intents.default()
        intents.members = True
        # Prefix commands read ordinary messages, which Discord gates behind the
        # privileged Message Content intent. Without it the bot still boots and
        # slash commands still work, but typed commands are invisible to it.
        intents.message_content = True
        super().__init__(
            command_prefix=config.prefix, intents=intents, help_command=None
        )

        self.db = Database(config.db_path)
        self.economy = EconomyService(self.db)
        self.market = MarketService(self.db, config.market)
        self.trading = TradingService(self.db, self.economy, self.market, config.trading)
        self._last_tick_at = 0.0
        self._known_guilds: set[int] = set()

    # -------------------------------------------------------------- lifecycle

    async def setup_hook(self) -> None:
        await self.db.connect()
        for extension in (
            "hera.cogs.economy",
            "hera.cogs.stocks",
            "hera.cogs.admin",
            "hera.cogs.help",
        ):
            await self.load_extension(extension)
            log.info("loaded extension %s", extension)

        if config.guild_id:
            guild = discord.Object(id=config.guild_id)
            self.tree.copy_global_to(guild=guild)
            synced = await self.tree.sync(guild=guild)
            log.info("synced %d commands to guild %s", len(synced), config.guild_id)
        else:
            synced = await self.tree.sync()
            log.info("synced %d global commands", len(synced))

        self.market_ticker.start()

    async def close(self) -> None:
        if self.market_ticker.is_running():
            self.market_ticker.cancel()
        await self.db.close()
        await super().close()

    async def on_ready(self) -> None:
        log.info("logged in as %s (%s)", self.user, self.user.id if self.user else "?")
        self._known_guilds = {guild.id for guild in self.guilds}
        await self.change_presence(
            activity=discord.Activity(
                type=discord.ActivityType.watching, name="the Hera exchange"
            )
        )

    async def on_guild_join(self, guild: discord.Guild) -> None:
        self._known_guilds.add(guild.id)

    # ------------------------------------------------------------ market ticks

    @tasks.loop(seconds=config.market.tick_seconds)
    async def market_ticker(self) -> None:
        """Advance every known guild's market once per tick."""
        # Refresh from the live guild list so markets for guilds joined while
        # offline are picked up without needing on_guild_join to have fired.
        self._known_guilds.update(guild.id for guild in self.guilds)
        for guild_id in list(self._known_guilds):
            try:
                await self.run_market_tick(guild_id)
            except Exception:  # pragma: no cover - defensive
                log.exception("market tick failed for guild %s", guild_id)

    @market_ticker.before_loop
    async def before_market_ticker(self) -> None:
        await self.wait_until_ready()

    async def run_market_tick(self, guild_id: int) -> None:
        """One full market tick: prices, dividends, orders, alerts, news."""
        result = await self.market.tick(guild_id)

        if result.dividends:
            await self.trading.apply_dividends(guild_id, result.dividends)

        fills = await self.trading.process_open_orders(guild_id)
        alerts = await self.trading.triggered_alerts(guild_id)
        margin_calls = await self.trading.margin_calls(guild_id)

        await self._announce(guild_id, result, fills, alerts, margin_calls)

    async def _announce(self, guild_id, result, fills, alerts, margin_calls) -> None:
        """Publish tick outcomes to the configured channel and to users."""
        if not result.news and not fills and not alerts and not margin_calls:
            return

        channel = self.get_channel(config.market_channel_id) if config.market_channel_id else None
        if isinstance(channel, discord.abc.Messageable):
            if result.news:
                headlines = "\n".join(
                    f"{'🟢' if item.impact >= 0 else '🔴'} {item.headline}"
                    for item in result.news[:6]
                )
                embed = discord.Embed(
                    title=f"📰 Market wire — tick {result.tick:,}",
                    description=headlines,
                    color=config.embed_color,
                )
                embed.set_footer(
                    text=f"Index {result.index_value:,.2f} ({percent(result.index_change * 100)})"
                )
                try:
                    await channel.send(embed=embed)
                except discord.HTTPException:
                    log.warning("could not post market news to channel %s", config.market_channel_id)

        # Order fills and alerts are direct messages: they are private by nature.
        for user_id, fill in fills:
            user = self.get_user(user_id)
            if user is None:
                continue
            embed = discord.Embed(
                title="🧾 Limit order filled",
                description=(
                    f"{fill.message}\n{signed(fill.realized_pnl)} realised" if fill.realized_pnl
                    else fill.message
                ),
                color=config.embed_color,
            )
            await self._safe_dm(user, embed)

        for alert in alerts:
            user = self.get_user(alert["user_id"])
            if user is None:
                continue
            embed = discord.Embed(
                title=f"🔔 {alert['symbol']} is {alert['direction']} {price(alert['threshold'])}",
                description=f"Last price: **{price(alert['price'])}**",
                color=config.embed_color,
            )
            if alert["note"]:
                embed.set_footer(text=alert["note"])
            await self._safe_dm(user, embed)

        for call in margin_calls:
            user = self.get_user(call["user_id"])
            if user is None:
                continue
            embed = discord.Embed(
                title=f"⚠️ Margin call on {call['symbol']}",
                description=(
                    f"Your short of {call['quantity']:,} {call['symbol']} is down "
                    f"{money(call['loss'])} against {money(call['collateral'])} of collateral.\n"
                    f"Entry {price(call['entry'])} → now {price(call['price'])}.\n"
                    "Cover the position or post more collateral."
                ),
                color=config.error_color,
            )
            await self._safe_dm(user, embed)

    async def _safe_dm(self, user: discord.User, embed: discord.Embed) -> None:
        try:
            await user.send(embed=embed)
        except (discord.Forbidden, discord.HTTPException):
            # DMs closed is normal; nothing actionable to do.
            pass

    # ---------------------------------------------------------------- errors

    async def on_app_command_error(
        self, interaction: discord.Interaction, error: app_commands.AppCommandError
    ) -> None:
        original = getattr(error, "original", error)
        if isinstance(original, HeraError):
            message = str(original)
        else:
            log.error("command error", exc_info=original)
            message = "Something went wrong handling that command."

        embed = discord.Embed(title=f"❌ {message}", color=config.error_color)
        try:
            if interaction.response.is_done():
                await interaction.followup.send(embed=embed, ephemeral=True)
            else:
                await interaction.response.send_message(embed=embed, ephemeral=True)
        except discord.HTTPException:
            pass

    async def on_command_error(
        self, ctx: commands.Context, error: commands.CommandError
    ) -> None:
        """Report prefixed-command failures the same way slash failures are shown."""
        if isinstance(error, commands.CommandNotFound):
            return
        if isinstance(error, commands.CommandInvokeError):
            original: BaseException = error.original
        else:
            original = error

        if isinstance(original, HeraError):
            message = str(original)
        elif isinstance(original, commands.UserInputError):
            message = _usage_error(ctx, error)
        else:
            log.error("prefix command error", exc_info=original)
            message = "Something went wrong handling that command."

        embed = discord.Embed(title=f"❌ {message}", color=config.error_color)
        try:
            await ctx.reply(embed=embed)
        except discord.HTTPException:
            pass

File: hera/bot.ts
/**
 * Bot wiring: shared services, command registration, background market ticker and
 * event dispatch.
 *
 * Commands are discovered from ``hera/commands/<category>/*.ts`` and registered on
 * both front ends from one definition. Prefix invocation is handled by a
 * ``MessageCreate`` listener; slash invocation by the interaction handler; both
 * funnel through the same handler in :mod:`hera/framework/register`.
 */
import { join } from 'node:path';
import {
  ActivityType,
  Client,
  Events,
  GatewayIntentBits,
  MessageFlags,
  Partials,
  REST,
  Routes,
  type AutocompleteInteraction,
  type ChatInputCommandInteraction,
  type Message,
  type User,
} from 'discord.js';
import { config } from './config';
import { Database } from './database';
import { HeraError } from './errors';
import { loadCommands } from './framework/registry';
import {
  dispatchPrefix,
  dispatchSlash,
  handleAutocomplete,
  matchPrefix,
  usage,
  PrefixArgumentError,
  type CommandHandler,
  type PrefixCommand,
} from './framework/register';
import type { CommandDefinition, CommandServices } from './framework/types';
import { EconomyService } from './services/economy';
import { MarketService } from './services/market';
import { TradingService } from './services/trading';
import { GamblingService } from './gambling/service';
import { errorEmbed } from './ui/embeds';
import { ViewRegistry } from './ui/views';
import { registerEvents, type EventContext } from './events';
import { outgoingFor } from './events/marketevents';

export type Logger = (message: string, error?: unknown) => void;

export interface BotOptions {
  log?: (message: string) => void;
  logError?: (message: string, error?: unknown) => void;
}

export class HeraBot extends Client {
  readonly db: Database;
  readonly economy: EconomyService;
  readonly market: MarketService;
  readonly trading: TradingService;
  readonly gambling: GamblingService;
  readonly commands = new Map<string, CommandDefinition>();
  readonly services: CommandServices;
  readonly ownerId: string | null = config.ownerId;
  readonly views = new ViewRegistry();

  private readonly definitions: CommandDefinition[] = [];
  private ticker: NodeJS.Timeout | null = null;
  private readonly knownGuilds = new Set<string>();
  private readonly prefixCommands = new Map<string, PrefixCommand>();
  private readonly displayNames = new Map<string, string>();
  private started = false;

  private readonly botOptions: BotOptions;

  constructor(options: BotOptions = {}, dbPath?: string) {
    super({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMembers,
        // Two intents gate prefix commands. GuildMessages is what makes Discord
        // deliver MESSAGE_CREATE for guild channels at all; without it the bot
        // still boots and slash commands still work, but typed commands never
        // arrive. MessageContent is the privileged intent that fills in the
        // message text once an event is delivered.
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
        GatewayIntentBits.DirectMessages,
      ],
      partials: [Partials.Channel],
    });
    this.botOptions = options;

    this.db = new Database(dbPath ?? config.dbPath);
    this.economy = new EconomyService(this.db);
    this.market = new MarketService(this.db, config.market);
    this.trading = new TradingService(this.db, this.economy, this.market, config.trading);
    this.gambling = new GamblingService(this.db, this.economy, config.gambling);

    this.services = {
      market: this.market,
      trading: this.trading,
      economy: this.economy,
      gambling: this.gambling,
      displayName: (userId) => this.displayNames.get(userId) ?? `User ${userId}`,
      runMarketTick: (guildId) => this.runMarketTick(guildId),
      registerView: (messageId, paginator) => this.views.register(messageId, paginator),
      commandIndex: { commands: this.commands },
      ownerId: this.ownerId,
    };
  }

  // -------------------------------------------------------------- lifecycle

  /** Discover commands and wire the gateway listeners. Idempotent. */
  async start(): Promise<void> {
    if (this.started) return;
    this.started = true;

    const { commands, failures } = loadCommands(join(__dirname, 'commands'));
    for (const [path, error] of failures) {
      this.botOptions.logError?.(`failed to load command file ${path}`, error);
    }
    for (const definition of commands) {
      if (this.commands.has(definition.name)) {
        this.botOptions.logError?.(`duplicate command name: ${definition.name}`);
        continue;
      }
      this.definitions.push(definition);
      this.commands.set(definition.name, definition);
      this.prefixCommands.set(definition.name, { definition, handler: this.handlerFor(definition) });
    }
    this.botOptions.log?.(`loaded ${this.definitions.length} commands`);

    registerEvents(this, this.eventContext());

    this.once(Events.ClientReady, () => void this.onReady());
    this.on(Events.InteractionCreate, (interaction) => void this.onInteraction(interaction));
    this.on(Events.MessageCreate, (message) => void this.onMessage(message));

    await this.login(config.token);
    await this.syncCommands();
  }

  private handlerFor(definition: CommandDefinition): CommandHandler {
    return (ctx, args) => definition.execute(ctx, args, this.services);
  }

  private async onReady(): Promise<void> {
    this.botOptions.log?.(`logged in as ${this.user?.tag ?? '?'}`);
    for (const guild of this.guilds.cache.values()) this.knownGuilds.add(guild.id);
    await this.startTicker();
    void this.user?.setActivity({ name: 'the Hera exchange', type: ActivityType.Watching });
  }

  private eventContext(): EventContext {
    return {
      guildJoined: (guildId) => this.knownGuilds.add(guildId),
      ensureGuild: (guildId) => this.knownGuilds.add(guildId),
      onTick: (guildId, result) =>
        this.announceTick(guildId, result as import('./market/engine').TickResult),
    };
  }

  // ---------------------------------------------------------- registration

  private async syncCommands(): Promise<void> {
    const rest = new REST().setToken(config.token);
    const payloads = this.definitions.map((definition) => buildCommandPayload(definition));

    if (config.guildId) {
      await rest.put(Routes.applicationGuildCommands(this.user!.id, config.guildId), {
        body: payloads,
      });
      this.botOptions.log?.(`synced ${payloads.length} commands to guild ${config.guildId}`);
    } else {
      await rest.put(Routes.applicationCommands(this.user!.id), { body: payloads });
      this.botOptions.log?.(`synced ${payloads.length} global commands`);
    }
  }

  // ---------------------------------------------------------------- ticker

  private async startTicker(): Promise<void> {
    if (this.ticker) return;
    const intervalMs = Math.max(1, config.market.tickSeconds) * 1000;
    this.ticker = setInterval(() => {
      void this.tickAllGuilds();
    }, intervalMs);
    // Do not hold the process open solely for the ticker.
    this.ticker.unref?.();
    if (config.runMarketOnStartup) await this.tickAllGuilds();
  }

  private async tickAllGuilds(): Promise<void> {
    for (const guild of this.guilds.cache.values()) this.knownGuilds.add(guild.id);
    for (const guildId of [...this.knownGuilds]) {
      try {
        await this.runMarketTick(guildId);
      } catch (error) {
        this.botOptions.logError?.(`market tick failed for guild ${guildId}`, error);
      }
    }
  }

  // ------------------------------------------------------------ interactions

  private async onInteraction(interaction: import('discord.js').Interaction): Promise<void> {
    if (await this.views.dispatch(interaction)) return;
    if (interaction.isAutocomplete()) {
      await this.onAutocomplete(interaction);
      return;
    }
    if (interaction.isChatInputCommand()) {
      await this.onSlashCommand(interaction);
    }
  }

  private async onAutocomplete(interaction: AutocompleteInteraction): Promise<void> {
    const definition = this.commands.get(interaction.commandName);
    if (!definition) {
      await interaction.respond([]);
      return;
    }
    try {
      await handleAutocomplete(definition, interaction, this.services);
    } catch {
      await interaction.respond([]);
    }
  }

  private async onSlashCommand(interaction: ChatInputCommandInteraction): Promise<void> {
    const definition = this.commands.get(interaction.commandName);
    if (!definition) return;
    try {
      await dispatchSlash(definition, this.handlerFor(definition), interaction);
    } catch (error) {
      await this.reportSlashError(interaction, error);
    }
  }

  private async reportSlashError(
    interaction: ChatInputCommandInteraction,
    error: unknown,
  ): Promise<void> {
    const message = this.describeError(error);
    const embed = errorEmbed(message);
    try {
      if (interaction.deferred || interaction.replied) {
        await interaction.followUp({ embeds: [embed], flags: MessageFlags.Ephemeral });
      } else {
        await interaction.reply({ embeds: [embed], flags: MessageFlags.Ephemeral });
      }
    } catch {
      // The interaction token can expire; nothing actionable to do.
    }
  }

  // -------------------------------------------------------------- prefix

  private async onMessage(message: Message): Promise<void> {
    if (message.author.bot) return;
    if (!this.user) return;
    const matched = matchPrefix(message.content, config.prefix, this.prefixCommands);
    if (!matched) return;
    if (message.guild) this.knownGuilds.add(message.guild.id);
    this.cacheUser(message.author);
    try {
      await dispatchPrefix(matched.command.definition, matched.command.handler, message, matched.tokens);
    } catch (error) {
      await this.reportPrefixError(message, matched.command.definition, error);
    }
  }

  private async reportPrefixError(
    message: Message,
    definition: CommandDefinition,
    error: unknown,
  ): Promise<void> {
    let text: string;
    if (error instanceof HeraError) {
      text = error.message;
    } else if (error instanceof PrefixArgumentError) {
      text = error.message;
    } else {
      this.botOptions.logError?.('prefix command error', error);
      text = 'Something went wrong handling that command.';
    }
    const embed =
      text.includes(config.prefix) && !(error instanceof HeraError)
        ? errorEmbed(text)
        : errorEmbed(text + (error instanceof HeraError ? `\n${usage(definition, config.prefix)}` : ''));
    try {
      await message.reply({ embeds: [embed] });
    } catch {
      // The message may have been deleted mid-command.
    }
  }

  private describeError(error: unknown): string {
    if (error instanceof HeraError) return error.message;
    this.botOptions.logError?.('command error', error);
    return 'Something went wrong handling that command.';
  }

  // ------------------------------------------------------------ market tick

  /** One full market tick: prices, dividends, orders, alerts, news. */
  async runMarketTick(guildId: string): Promise<void> {
    const result = this.market.tick(guildId);

    if (Object.keys(result.dividends).length > 0) {
      this.trading.applyDividends(guildId, result.dividends);
    }

    const fills = this.trading.processOpenOrders(guildId);
    const alerts = this.trading.triggeredAlerts(guildId);
    const marginCalls = this.trading.marginCalls(guildId);

    await this.announceTick(guildId, result, fills, alerts, marginCalls);
  }

  private async announceTick(
    guildId: string,
    result: import('./market/engine').TickResult,
    fills: [string, import('./services/trading').Fill][] = [],
    alerts: import('./services/trading').AlertHit[] = [],
    marginCalls: import('./services/trading').MarginCall[] = [],
  ): Promise<void> {
    void guildId;
    const messages = outgoingFor(result, fills, alerts, marginCalls);
    for (const message of messages) {
      if (message.channel) {
        await this.postToMarketChannel(message.embed);
      } else if (message.userId) {
        // Order fills and alerts are private, so they go out as direct messages.
        await this.safeDm(message.userId, message.embed);
      }
    }
  }

  private async postToMarketChannel(embed: import('discord.js').EmbedBuilder): Promise<void> {
    if (!config.marketChannelId) return;
    const channel = this.channels.cache.get(config.marketChannelId);
    if (!channel || !('send' in channel)) return;
    try {
      await channel.send({ embeds: [embed] });
    } catch {
      this.botOptions.logError?.(`could not post market news to ${config.marketChannelId}`);
    }
  }

  private async safeDm(userId: string, embed: import('discord.js').EmbedBuilder): Promise<void> {
    try {
      const user = await this.users.fetch(userId);
      await user.send({ embeds: [embed] });
    } catch {
      // DMs closed is normal; nothing actionable to do.
    }
  }

  cacheUser(user: User): void {
    this.displayNames.set(user.id, user.displayName);
  }

  // ---------------------------------------------------------------- shutdown

  async shutdown(): Promise<void> {
    if (this.ticker) clearInterval(this.ticker);
    this.ticker = null;
    this.db.close();
    await this.destroy();
  }
}

// ------------------------------------------------------------- payload helpers

function toOptionPayload(arg: import('./framework/types').CommandArgument): Record<string, unknown> {
  const base: Record<string, unknown> = {
    name: arg.name,
    description: arg.description.slice(0, 100),
    required: arg.required ?? false,
  };
  switch (arg.type) {
    case 'string':
      base.type = 3;
      if (arg.choices) base.choices = arg.choices;
      if (arg.autocomplete) base.autocomplete = true;
      break;
    case 'integer':
      base.type = 4;
      if (arg.minValue !== undefined) base.min_value = arg.minValue;
      if (arg.maxValue !== undefined) base.max_value = arg.maxValue;
      break;
    case 'number':
      base.type = 10;
      if (arg.minValue !== undefined) base.min_value = arg.minValue;
      if (arg.maxValue !== undefined) base.max_value = arg.maxValue;
      break;
    case 'user':
      base.type = 6;
      break;
  }
  return base;
}

function buildCommandPayload(definition: CommandDefinition): Record<string, unknown> {
  return {
    name: definition.name,
    description: definition.description.slice(0, 100),
    options: (definition.args ?? []).map(toOptionPayload),
  };
}
